File: js/p039.js
"use strict";
var tools = require("./util/tools"),
    pythagoreanTriple = tools.pythagoreanTriple,
    gcd = tools.gcd;
/*
 * Integer right triangles
 * Problem 39
 * If p is the perimeter of a right angle triangle with integral length 
 * sides, {a,b,c}, there are exactly three solutions for p = 120.
 * 
 * {20,48,52}, {24,45,51}, {30,40,50}
 * 
 * For which value of p ≤ 1000, is the number of solutions maximised?
 * 
 * Answer: 840
 */

function init() {
    var limit = 1000;
    var perimeters = [];
    for (var i=0; i<=limit; i++) {
        perimeters.push(0);
    }
    
    for (var m=2; m<1000; m++) {
        for (var n=1; n<m; n++) {
            if (gcd(m, n) == 1 && (isEven(m) || isEven(n)) ) {
                var t = pythagoreanTriple(m, n)
                var p = t.a + t.b + t.c;
                var k = 1;
                while (p*k <= limit) {
                    perimeters[p*k] += 1;
                    k += 1;
                }    
            }          
        }
    }

    var max = 0;
    var output = 0;
    for (var i=0; i<=limit; i++) {
        if (perimeters[i] > max) {
            max = perimeters[i];
            output = i;
        }
    }

    return output;
}

function isEven(n) {
    return n % 2 == 0;
}

module.exports.init = init;

File: python/p039.py
"""
	
Project Euler - Problem 39 - Integer right triangles
https://projecteuler.net/problem=39

"""
import timing
from math import gcd
from eulerutil import pythagorean_triple, odd


def main():
    limit = 1001
    freq = {}

    n = 2
    while True:
        for m in range(1, n):
            if gcd(m, n) != 1 or (odd(m) and odd(n)): continue 
            perim = sum(pythagorean_triple(m, n))
            
            if perim > limit:
                n += 1
                break

            k = 1
            while True:
                if k * perim > limit: break 
                freq[perim*k] = freq.get(perim*k, 0) + 1
                k += 1
        else:
            n += 1
        if sum(pythagorean_triple(1, n)) > limit or n > 30:
            break

    print(max(freq, key=lambda x: freq[x]))


if __name__ == '__main__':
	main()
